gio/src/list_model.rs

```diff
@@ -1,33 +1,214 @@
 // Take a look at the license at the top of the repository in the LICENSE file.
 
 use crate::prelude::*;
-use crate::ListModel;
+use glib::object::Cast;
+use glib::object::IsA;
+use glib::signal::connect_raw;
+use glib::signal::SignalHandlerId;
+use glib::translate::*;
+use glib::Object;
+use std::boxed::Box as Box_;
+use std::mem::transmute;
+
+glib::wrapper! {
+    pub struct ListModel<T: (IsA<Object>)>(Interface<ffi::GListModel, ffi::GListModelInterface>),
+        @requires_generic
+            <Super: (IsA<Super>) + (IsA<Object>), Sub: (IsA<Super>) + (IsA<Object>)> ListModel<Super> for ListModel<Sub>,
+        @default_casts false,
+        @checkers ListModelCastChecker<T>, ListModelValueChecker<T>;
+
+    match fn {
+        type_ => || ffi::g_list_model_get_type(),
+    }
+}
+
+#[doc(hidden)]
+pub struct ListModelCastChecker<T>(std::marker::PhantomData<T>);
+
+impl<T: IsA<Object>> glib::object::ObjectCastChecker<ListModel<T>> for ListModelCastChecker<T> {
+    fn check<U: glib::ObjectType>(obj: &U) -> bool {
+        if glib::object::GenericObjectCastChecker::<ListModel<T>>::check(obj) {
+            let item_type: glib::Type =
+                unsafe { from_glib(ffi::g_list_model_get_item_type(obj.as_ptr() as *mut _)) };
+            if item_type.is_a(T::static_type()) {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+#[doc(hidden)]
+pub struct ListModelValueChecker<T>(std::marker::PhantomData<T>);
+
+unsafe impl<T: IsA<Object>> glib::value::ValueTypeChecker for ListModelValueChecker<T> {
+    type Error = glib::value::ValueTypeMismatchOrNoneError<glib::value::ValueTypeMismatchError>;
+
+    fn check(value: &glib::Value) -> Result<(), Self::Error> {
+        glib::object::ObjectValueTypeChecker::<ListModel<Object>>::check(value)?;
+        let model: &ListModel<Object> = unsafe { glib::value::FromValue::from_value(value) };
+        let model_type = model.item_type();
+        let expected = T::static_type();
+        if !model_type.is_a(expected) {
+            return Err(glib::value::ValueTypeMismatchError::new(model_type, expected).into());
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: IsA<Object>> ListModel<T> {
+    pub const NONE: Option<&'static ListModel<T>> = None;
+}
+
+impl<T: IsA<Object>> std::fmt::Display for ListModel<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str("ListModel")
+    }
+}
+
+pub trait ListModelExt: 'static {
+    #[doc(alias = "g_list_model_get_item_type")]
+    #[doc(alias = "get_item_type")]
+    fn item_type(&self) -> glib::types::Type;
+
+    #[doc(alias = "g_list_model_get_n_items")]
+    #[doc(alias = "get_n_items")]
+    fn n_items(&self) -> u32;
+
+    #[doc(alias = "g_list_model_get_object")]
+    #[doc(alias = "get_object")]
+    fn object(&self, position: u32) -> Option<Object>;
```

Comment:
There isn't any real difference in the C API besides the return type. I think `fn object` can just be removed.

Returning `T` is inconvenient because then you have to put `<T>` on `ListModelExt` and do `ListModelExt::<Type>::object(&model, 123)` if you want to use the turbofish operator. Putting the parameter on `fn item` lets you do `model.item::<Type>(123)`

Reply:
The reason why `get_item()` and `get_object()` are separate is because `get_item()` returns a `void*`, whereas `get_object()` returns a `GObject*`. The initial design of GListModel was made generic enough to support any GType payload that could fit into a pointer. Of course, it's completely unbindable *ex post*, but every now and again somebody thinks they can outsmart every binding in existence.

Reply:
Then perhaps `fn object` should be kept, and `item` removed.

Reply:
i don't think so, because bindings are supposed to use get_object internally and rename it to get_item; to be consistent with get_n_items